### ISSUE TITLE
gprc-node can prevent node from exiting.

### ISSUE THREAD

**Issue by mikelehen** [grpc/grpc#8083]
I've noticed that certain error conditions seem to leave node completely hung instead of exiting cleanly.  node stops executing code and ctrl-C or "kill <pid>" don't work.  I have to "kill -9 <pid>" to get the process to die.

I don't have a minimal repro, but for instance I seem to be able to reproduce it by trying to send a message over a stream that doesn't conform to the proto.  It triggers an exception (below) but instead of node exiting cleanly, it just hangs.

```
Error: Illegal value for Message.Field [redacted] of type bytes: object (proto3 field without field presence cannot be null)
    at Error (native)
    at fail (/usr/local/google/home/mikelehen/tmp/test/node_modules/protobufjs/dist/protobuf.js:3440:23)
    at T.ProtoBuf.Reflect.FieldPrototype.verifyValue (/usr/local/google/home/mikelehen/tmp/test/node_modules/protobufjs/dist/protobuf.js:3446:21)
    at MessagePrototype.set (/usr/local/google/home/mikelehen/tmp/test/node_modules/protobufjs/dist/protobuf.js:2475:59)
    at MessagePrototype.set (/usr/local/google/home/mikelehen/tmp/test/node_modules/protobufjs/dist/protobuf.js:2466:38)
    at Message (/usr/local/google/home/mikelehen/tmp/test/node_modules/protobufjs/dist/protobuf.js:2395:34)
    at serialize (/usr/local/google/home/mikelehen/tmp/test/node_modules/grpc/src/node/src/common.js:87:23)
    at ClientDuplexStream.serialize (/usr/local/google/home/mikelehen/tmp/test/node_modules/grpc/src/node/src/common.js:126:12)
    at ClientDuplexStream._write (/usr/local/google/home/mikelehen/tmp/test/node_modules/grpc/src/node/src/client.js:102:22)
    at doWrite (_stream_writable.js:307:12)

```

I did a strace on node, and it looks to my untrained eye like node is waiting for a thread to exit...

```
clone(child_stack=0x7fc4eb7fdf70, flags=CLONE_VM|CLONE_FS|CLONE_FILES|CLONE_SIGHAND|CLONE_THREAD|CLONE_SYSVSEM|CLONE_SETTLS|CLONE_PARENT_SETTID|CLONE_CHILD_CLEARTID, parent_tidptr=0x7fc4eb7fe9d0, tls=0x7fc4eb7fe700, child_tidptr=0x7fc4eb7fe9d0) = 4618
... much later, the last thing strace outputs is ...
futex(0x7fc4eb7fe9d0, FUTEX_WAIT, 4618, NULL
```

**Comment by murgatroid99** [comment]
I think the problem here is that the call should be ended with an error on the local side when a write fails. But since the Node code currently doesn't do that, grpc ends up waiting forever for the call to finish.

**Comment by vjpai** [comment]
@murgatroid99 : should this bug remain open?

**Comment by grosskur** [comment]
I'm seeing similar hangs with grpc 1.0.0. Basically, in situations where node would normally coredump, it now hangs indefinitely if the `grpc` module has been imported. However, this seems to be fixed in grpc 1.0.1 (released on 2016-10-27).

@mikelehen: Can you verify if the hang you saw is fixed with grpc 1.0.1, too?

@murgatroid99: Do you know which specific commits would have fixed this?

Here's simple way to reproduce it. Run this command and observe a `RangeError` crash:

```
$ node -e 'require("grpc"); function crash() { crash(); } crash();'
[eval]:1
require("grpc"); function crash() { crash(); } crash();
                               ^

RangeError: Maximum call stack size exceeded
    at crash ([eval]:1:32)
    at crash ([eval]:1:37)
    at crash ([eval]:1:37)
    at crash ([eval]:1:37)
    at crash ([eval]:1:37)
    at crash ([eval]:1:37)
    at crash ([eval]:1:37)
    at crash ([eval]:1:37)
    at crash ([eval]:1:37)
    at crash ([eval]:1:37)
```

With grpc 1.0.0, adding `--abort-on-uncaught-exception` results in a hang:

```
$ node --abort-on-uncaught-exception -e 'require("grpc"); function crash() { crash(); } crash();'
<error>

FROM
```

We know this is caused by grpc because removing `require("grpc")` results in a coredump as expected:

```
$ node --abort-on-uncaught-exception -e 'function crash() { crash(); } crash();'
<error>

FROM
Illegal instruction (core dumped)
```

With grpc 1.0.1, `--abort-on-uncaught-exception` results in a coredump as expected:

```
$ node --abort-on-uncaught-exception -e 'require("grpc"); function crash() { crash(); } crash();'
<error>

FROM
Illegal instruction (core dumped)
```

I've verified this with both node 4.7.0 and 6.9.2. (I couldn't get grpc to compile with 7.3.0.)

**Comment by murgatroid99** [comment]
I mentioned this in an email: I don't know what changes in 1.0.1 could have caused a change in this behavior. Since this bug appears to no longer be occurring, I will close this issue.
